### PR TITLE
ArangoSearch advanced AQL functions

### PR DESCRIPTION
**Pull request by gnusi**
Please note that for legal reasons we require you to sign the 
[Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/174/

File: Documentation/Books/AQL/Views/ArangoSearch.md [hunk 1]
```diff
@@ -59,6 +59,32 @@ operators, e.g.:
...
+
+### ANALYZER()
+
+`ANALYZER(filter-expression, analyzer)`
+
+Override analyzer in a context of **filter-expression** with another one, denoted
+by a specified **analyzer** argument, making it available for filter functions.
+
+- *filter-expression* - any valid filter expression
+- *analyzer* - string with the analyzer to inmbue, i.e. *"text_en"* or [one of the other available string analyzers](../../../Manual/Views/ArangoSearch/Analyzers.html)
```
Comment (jsteemann) [inline Documentation/Books/AQL/Views/ArangoSearch.md]: inmbue -> imbue

File: Documentation/Books/AQL/Views/ArangoSearch.md [hunk 2]
```diff
@@ -206,7 +282,7 @@ to watch the analyzer doing its work
 to match documents where 'description' best matches 'a quick brown fox'
 
     FOR doc IN VIEW someView
-      FILTER doc.description IN TOKENS('a quick brown fox', 'text_en')
+      FILTER ANALYZER(doc.description IN TOKENS('a quick brown fox', 'text_en'), 'test_en')
```
Comment (jsteemann) [inline Documentation/Books/AQL/Views/ArangoSearch.md]: should this be "test_en" or "text_en"?

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 3]
```diff
@@ -135,14 +223,16 @@ bool byTerm(
   switch (value.type()) {
     case arangodb::iresearch::SCOPED_VALUE_TYPE_NULL:
       if (filter) {
-        arangodb::iresearch::kludge::mangleNull(name);
+        kludge::mangleNull(name);
         filter->field(std::move(name));
+        filter->boost(filterCtx.boost);
         filter->term(irs::null_token_stream::value_null());
       } return true;
```
Comment (jsteemann) [inline arangod/IResearch/IResearchFilterFactory.cpp]: can you put the `return true` on its own line here for better readability?

Comment (jsteemann) [reply arangod/IResearch/IResearchFilterFactory.cpp]: (same a bit further below)

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 4]
```diff
@@ -1070,250 +1373,254 @@ bool fromFuncExists(
...
-
-        if (!analyzerId.execute(ctx)) {
-          LOG_TOPIC(WARN, arangodb::iresearch::TOPIC)
-            << "'EXISTS' AQL function: Failed to evaluate 3rd argument";
-          return false;
-        }
-
-        if (!analyzerId.getString(arg)) {
+      if (argc > 2) {
+        if (TypeAnalyzer.c_str() != typeHandler->first.c_str()) {
```
Comment (jsteemann) [inline arangod/IResearch/IResearchFilterFactory.cpp]: is this intended as a string comparion?
if not, can we make sure the string pointers will always be the same?

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 5]
```diff
@@ -1070,250 +1373,254 @@ bool fromFuncExists(
...
-    args.getMemberUnchecked(0), *ctx.ref
-  );
+  auto const lastArg = argc - 1;
+  auto minMatchCountArg = args.getMemberUnchecked(lastArg);
 
-  if (!fieldArg) {
+  if (!minMatchCountArg) {
     LOG_TOPIC(WARN, arangodb::iresearch::TOPIC)
-      << "'PHRASE' AQL function: 1st argument is invalid";
+      << "'ANALYZE' AQL function: " << lastArg << " argument is invalid";
```
Comment (jsteemann) [inline arangod/IResearch/IResearchFilterFactory.cpp]: 'ANALYZE' => 'MIN_MATCH'

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 6]
```diff
@@ -1070,250 +1373,254 @@ bool fromFuncExists(
...
-  // last argument defines the analyzer to use
-  // ...........................................................................
-
-  auto const* analyzerArg = args.getMemberUnchecked(argc - 1);
-
-  if (!analyzerArg) {
+  if (!minMatchCountArg->isDeterministic()) {
     LOG_TOPIC(WARN, arangodb::iresearch::TOPIC)
-      << "'PHRASE' AQL function: Unable to parse analyzer value";
+      << "'ANALYZE' AQL function: " << lastArg << " argument is intended to be deterministic";
```
Comment (jsteemann) [inline arangod/IResearch/IResearchFilterFactory.cpp]: 'ANALYZE' => 'MIN_MATCH'

File: arangod/IResearch/IResearchFilterFactory.cpp (hunk 4 above)
Comment (gnusi) [reply arangod/IResearch/IResearchFilterFactory.cpp]: that is intentional pointer comparison.
`irs::string_ref const TypeAnalyzer("analyzer");` is defined above and will contain same `const char*`

File: Documentation/Books/AQL/Views/ArangoSearch.md (hunk 1 above)
Comment (gnusi) [reply Documentation/Books/AQL/Views/ArangoSearch.md]: done

File: Documentation/Books/AQL/Views/ArangoSearch.md (hunk 2 above)
Comment (gnusi) [reply Documentation/Books/AQL/Views/ArangoSearch.md]: done

File: arangod/IResearch/IResearchFilterFactory.cpp (hunk 4 above)
Comment (jsteemann) [reply arangod/IResearch/IResearchFilterFactory.cpp]: ok

**Comment by jsteemann** [comment]
Started another test:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/180/

**Comment by gnusi** [comment]
passed
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/182/

**Pull request merged**